src/main/scala/org/constellation/infrastructure/endpoints/CheckpointEndpoints.scala

```diff
@@ -54,33 +69,64 @@ class CheckpointEndpoints[F[_]](implicit F: Concurrent[F], C: ContextShift[F]) e
-  ): HttpRoutes[F] = HttpRoutes.of[F] {
-    case req @ POST -> Root / "finished" / "checkpoint" =>
-      for {
-        finishedCheckpoint <- req.decodeJson[FinishedCheckpoint]
-        baseHash = finishedCheckpoint.checkpointCacheData.checkpointBlock.baseHash
-        _ <- logger.debug(s"Handle finished checkpoint for cb: $baseHash")
-        _ <- metrics.incrementMetricAsync[F]("peerApiRXFinishedCheckpoint")
-        accept = checkpointAcceptanceService.acceptWithNodeCheck(finishedCheckpoint)
-        response <- snapshotService.getNextHeightInterval.flatMap { nextHeight =>
-          (nextHeight, finishedCheckpoint.checkpointCacheData.height) match {
-            case (_, None) =>
-              logger.warn(s"Missing height when accepting block hash=$baseHash") >> BadRequest()
-            case (2, _) =>
-              F.start(accept) >> Accepted()
-            case (nextHeight, Some(Height(min, _))) if nextHeight > min =>
-              logger.debug(
-                s"Handle finished checkpoint for cb: $baseHash height condition not met next interval: $nextHeight received: $min"
-              ) >> Conflict()
-            case (_, _) =>
-              F.start(accept) >> Accepted()
+  ): HttpRoutes[F] =
+    HttpRoutes.of[F] {
+      case req @ POST -> Root / "peer" / "checkpoint" / "finished" =>
+        for {
+          message <- req.as[GossipMessage[CheckpointBlockPayload]]
+          payload = message.payload
+          senderId = req.headers.get(`X-Id`).map(_.value).map(Id(_)).get
+
+          res <- messageValidator.validateForForward(message, senderId) match {
+            case Invalid(EndOfCycle)                                                            => checkpointBlockGossipService.finishCycle(message) >> Ok()
+            case Invalid(IncorrectReceiverId(_, _)) | Invalid(PathDoesNotStartAndEndWithOrigin) => BadRequest()
+            case Invalid(IncorrectSenderId(_))                                                  => Response[F](status = Unauthorized).pure[F]
+            case Invalid(e)                                                                     => logger.error(e)(e.getMessage) >> InternalServerError()
+            case Valid(_) =>
+              val accept = checkpointAcceptanceService.acceptWithNodeCheck(payload.block.value)
+              val processFinishedCheckpointAsync = F.start(
+                C.shift >>
+                  snapshotService.getNextHeightInterval.flatMap { nextHeight =>
+                    (nextHeight, payload.block.value.checkpointCacheData.height) match {
+                      case (_, None)                                              => F.unit
+                      case (2, _)                                                 => F.start(accept) >> F.unit
+                      case (nextHeight, Some(Height(min, _))) if nextHeight > min => F.unit
+                      case (_, _)                                                 => F.start(accept) >> F.unit
```

Comment:
If accept fails and I guess that's possible because acceptWithNodeCheck can throw, the spreading message will fail as well.